src/optimizer/rule/date_trunc_simplification.cpp

```diff
@@ -0,0 +1,483 @@
+					expr.left = column_part.Copy();
+					expr.right = std::move(trunc);
+				} else {
+					expr.right = column_part.Copy();
+					expr.left = std::move(trunc);
+				}
+			} else {
+				// If the RHS is already truncated (i.e.  date_trunc(part, rhs) = rhs), then we can use
+				// it as-is.
+				if (col_is_lhs) {
+					expr.left = column_part.Copy();
+					// Determine whether the RHS needs to be casted.
+					if (rhs.return_type.id() != expr.left->return_type.id()) {
+						expr.right = CastAndEvaluate(std::move(expr.right), expr.left->return_type);
+					}
+				} else {
+					expr.right = column_part.Copy();
+					// Determine whether the RHS needs to be casted.
+					if (rhs.return_type.id() != expr.right->return_type.id()) {
+						expr.left = CastAndEvaluate(std::move(expr.left), expr.right->return_type);
+					}
+				}
+			}
+
+			changes_made = true;
+			return nullptr;
+		}
+
+	case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+	case ExpressionType::COMPARE_GREATERTHAN:
+		// date_trunc(part, column) <= constant_rhs  -->  column <= date_trunc(part, date_add(constant_rhs,
+		//                                                                                    INTERVAL 1 part))
+		// date_trunc(part, column) >  constant_rhs  -->  column >= date_trunc(part, date_add(constant_rhs,
+		//                                                                                    INTERVAL 1 part))
+		{
+			// Create date_trunc(part, date_add(rhs, INTERVAL 1 part)) and fold the constant.
+			auto trunc = CreateTruncAdd(date_part, rhs, column_part.return_type);
+			if (!trunc) {
+				return nullptr; // Something went wrong---don't do the optimization.
+			}
+
+			if (col_is_lhs) {
+				expr.left = column_part.Copy();
+				expr.right = std::move(trunc);
+			} else {
+				expr.right = column_part.Copy();
+				expr.left = std::move(trunc);
+			}
+
+			// If this is a >, we need to change it to >= for correctness.
+			if (rhs_comparison_type == ExpressionType::COMPARE_GREATERTHAN) {
+				if (col_is_lhs) {
+					expr.SetExpressionTypeUnsafe(ExpressionType::COMPARE_GREATERTHANOREQUALTO);
+				} else {
+					expr.SetExpressionTypeUnsafe(ExpressionType::COMPARE_LESSTHANOREQUALTO);
+				}
+			}
+
+			changes_made = true;
+			return nullptr;
+		}
+
+	case ExpressionType::COMPARE_NOT_DISTINCT_FROM:
+		// date_trunc(part, column) IS NOT DISTINCT FROM constant_rhs
+		//
+		//   Here we have two cases: when constant_rhs is NULL, this simplifies to:
+		//
+		// column IS NULL
+		//
+		//   Otherwise, the expression becomes:
+		//
+		// (column >= date_trunc(part, constant_rhs) AND
+		//  column < date_trunc(part, date_add(constant_rhs, INTERVAL 1 part)) AND
+		//  column IS NOT NULL)
+		//
+		{
+			if (rhs.value.IsNull()) {
+				// Return 'column IS NULL'.
+				auto op = make_uniq<BoundOperatorExpression>(ExpressionType::OPERATOR_IS_NULL, LogicalType::BOOLEAN);
+				op->children.push_back(column_part.Copy());
+				return std::move(op);
+			} else {
```

Comment:
I'm wondering if we shouldn't fall through or share code with equality here?  Share the cases, but do the `NULL` check for IND.

Reply:
I thought about it when I implemented it and at the time I thought that it was cleaner for a reader to keep the optimizations separate, but that was a weak preference.  I merged them together in d95efbb68d.